### PR TITLE
feat: add HTTPClient implementation using curl as a module

### LINKED ISSUE #2903: Add HTTP client implementation using libcurl as a module, optionally replacing the default one
### Describe the project you are working on
3d multiplayer game

### Describe the problem or limitation you are having in your project
The current http api is terrible and is not able to be concurrent.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
To improve the http api, we can replace the current implementation with libcurl.

Libcurl supports HTTP/2 and any other versions as they're done.

This is already approved by a Networking team meeting.

Required features:

* Async file upload

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Implement libcurl similar to 
https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator/src/main/resources/C-libcurl.
 and https://github.com/OpenAPITools/openapi-generator/tree/master/samples/client/petstore/c

Use mbedtls: https://tls.mbed.org/kb/how-to/compile-curl-with-mbedtls
### If this enhancement will not be used often, can it be worked around with a few lines of script?
HTTP is not trivial.

### Is there a reason why this should be core and not an add-on in the asset library?
HTTP support is a core feature.

### PR DESCRIPTION
**Pull request by jonbonazza**
This PR uses libcurl in multi mode (curl_multi) to implement the HTTPClient interface. It cleans up the implementations significantly and should reduce maintenance burden while giving us easier access to more features in the end.

fixes https://github.com/godotengine/godot-proposals/issues/2903

## Scope

The goal of this PR is only to create a curl-based implementation of `HTTPClient` that conforms to the existing interface and nothing more.

The net team has discussed several improvements to the `HTTPClient` interface in general that may be implemented in future PRs, but they remain outside the scope of this one.

## Remaining Work

The current state of the PR is *DRAFT*. It still needs more testing and there are still a few remaining pieces left.

- [x] Support IPv6 DNS resolution using Godot's resolver APIs. I was unable to get this to work, but I consider it essential for this PR to be merged. IPv4 works fine though.
- [x] See if there's a way to make DNS resolution more performant by using the async apis. I'm not familiar enough with Godot's DNS APIs to know how best to use them, and the documentation for them is quite lacking. May want to just leave this up to a future improvement with its own PR.
- [x] Implement `set_connection` and `get_connection`. The way libcurl works doesn't lend itself well to the `StreamPeer` interface, so without some changes to that API, there's not a good way to do t his. If we _do_ decide to use `NetSocket` as the socket implementation (see above item) then we get this for free, so that's one pro to that approach. That said, I don't see these functions used anywhere in the Godot codebase, so I'd question whether they are even needed to begin with.
- [x] Get libcurl building on all platforms. Currently, I only have it working on Windows. I know @Faless has gotten it working on Linux, but I'm not sure what was necessary there. This might take some effort. We should also take the time to decide what features to enable/disable to tune binary size and also add a flag for using the system libcurl if available.

If I've missed anything in either the implementation itself, or the TODO list above, please don't hesitate to bring it up so we can discuss and potentially address it.

## Regarding Unit Tests

While we should certainly strive for as much unit tests coverage as possible, I didn't see any existing tests for the http client, so I've decided to leave that for an exercise for another day (and potentially another person). 

I want to stress that my decision not to include unit tests is not to the discredit of the idea of unit tests in general; quite the contrary, in fact. I would like to see a day where Godot's unit test coverage nears 100% just as much as anyone. However the best way to test http code is by running it against a real HTTP server, and to my knowledge, Godot doesn't currently have the facilities to do that and I'm not comfortable convoluting this particular PR by adding those facilities.

## A Note About The Commits

There's quite a lot of commits in this one and none of them are particularly clean. I'm a bit embarrassed about that, but it's mostly due to the fact I was learning so many new technologies along the way and ended up doing a lot of things wrong in a lot of different places in odd orders.

I suggest when merging this PR to squash all of the commits in order to keep things as clean as possible in master.

### CODE REVIEW

**Comment by Calinou** [comment]
> Currently, we only resolve one address for a domain. Godot's resolver APIs support resolving multiple

Wasn't this implemented by https://github.com/godotengine/godot/pull/49026?

**Comment by Faless** [comment]
Few things also mentioned in the chat:
- **Only fetch the body during `read_response_body_chunk`**. Not during poll :exclamation: 
- For IP resolution, see:
  https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/core/io/http_client_tcp.cpp#L81-L83
  https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/core/io/http_client_tcp.cpp#L250-L255


> * [blocking] I'm not even really sure how this is expected to be used. Is it even needed?

When you want to handle the request in a thread.

> * Implement `set_connection` and `get_connection`.

If we don't use our NetSocket implementation, just have them do nothing/return null and spit an error. Like JavaScript does:

https://github.com/godotengine/godot/blob/0caaaf4018feddb1d3e6e75870df282715efd276/platform/javascript/http_client_javascript.cpp#L74-L76

> * Docs... I dread this one.

Well, be happy, you won't have to write any new doc, given you are creating a new implementation of an exposed class, not a new exposed class (so HTTPClient is already fully documented).
You can of course try to improved the documentation, if you like :).

> * Get libcurl building on all platforms.

Yeah, this is going to be quite annoying, on Linux I only got it to work with system-wide mbedtls (not with built-in) and using @fire `curl_config.h`.

**Comment by Faless** [comment]
> While we should certainly strive for as much unit tests coverage as possible, I didn't see any existing tests for the http client, so I've decided to leave that for an exercise for another day (and potentially another person).

I had testing for HTTPClient on a separate repository like the one you are proposing (those are not unit tests):
https://github.com/Faless/gd-tests/

We can use those to verify the correct behaviour, the main problem was that websites would, from time to time slightly change their layout, so of course tests would unexpectedly break.

**Comment by Faless** [comment]
Size comparison (`target=release, builtin_mbedtls=no`):
Curl: `53436 KiB`
NoCurl: `53176 KiB`
Diff: `~260 KiB`

**Comment by jonbonazza** [comment]
> > Currently, we only resolve one address for a domain. Godot's resolver APIs support resolving multiple
> 
> Wasn't this implemented by #49026?
>

Yes, the engine supports it. This code, however does not. I am also unsure whether it's possible to do in an async manner. It's also worth mentioning that the current HTTPClient implementation does not seem to do multi address resolution, so Im gonna leave it to be outside the scope of this PR for now.

**Comment by jonbonazza** [comment]
Okay I've trimmed down the TODO list a bit.

> Only fetch the body during read_response_body_chunk. Not during poll ❗

Fixed this.

Also, @Faless I've updated the code to use the async resolver polling. It's basically the same code as the existing HTTPClientTCP you referenced, so nothing too crazy there.

Still can't get IPv6 to work and still not quite groking the blocking mode stuff. I've left a couple messages in the dev chat.

**Comment by jonbonazza** [comment]
This is ready for review.

I've tested this on Windows using the attached project, however it should probably be tested on linux, os, anxdroid, and ios as well for posterity.

[curl_test.zip](https://github.com/godotengine/godot/files/7071993/curl_test.zip)

File: modules/mbedtls_curl/http_client_curl.h
Comment (Faless) [inline modules/mbedtls_curl/http_client_curl.h:65]:
```suggestion
	int read_chunk_size = 65536;
```

**Comment by Faless** [comment]
Would be great to squash, leaving 2 commits, one adding the third-party library the other adding the module itself.

File: thirdparty/README.md
Comment (Faless) [inline thirdparty/README.md:41]: I think this is from a rebase issue?

File: thirdparty/README.md
Comment (Faless) [inline thirdparty/README.md:50]: we should mention our custom config too

**Comment by fire** [comment]
@jonbonazza Will work on rebasing into two prs and resolving the reviews.

File: thirdparty/README.md
Comment (fire) [reply thirdparty/README.md:41]: I resolved this issue because I think the original bundle is required for the fallback.

File: thirdparty/README.md
Comment (fire) [reply thirdparty/README.md:50]: Mentioned the derived config.

**Comment by fire** [comment]
Historical note: The module name `mbedtls_curl` was done to force this module to build after `mbedtls`.

**Comment by fire** [comment]
```bash
In file included from thirdparty/curl/lib/vtls/mbedtls.c:59:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
Compiling ==> thirdparty/curl/lib/bufref.c
1 error generated.
scons: *** [thirdparty/curl/lib/vtls/mbedtls.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/altsvc.c:35:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/altsvc.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/asyn-thread.c:67:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/asyn-thread.windows.opt.tools.64.o] Error 1
In file included from thirdparty/curl/lib/vtls/vtls.c:55:
thirdparty/curl/lib/sendf.h:37:2: error: "missing VARIADIC macro define, fix and rebuild!"
#error "missing VARIADIC macro define, fix and rebuild!"
 ^
1 error generated.
scons: *** [thirdparty/curl/lib/vtls/vtls.windows.opt.tools.64.o] Error 1
```

Using mingw llvm.

>  /bin/bash -c "PATH=/opt/llvm-mingw/bin:$PATH scons werror=no platform=windows target=release_debug -j`nproc` use_lto=no deprecated=no use_mingw=yes use_llvm=yes use_lld=yes use_thinlto=yes LINKFLAGS=-Wl,-pdb= CCFLAGS='-g -gcodeview' debug_symbols=no custom_modules=../godot_custom_modules"

**Comment by fire** [comment]
Discussed with Lyuma this might be a VARIADIC define missing. Can get a rebase in near future.

**Comment by fire** [comment]
@Faless @jonbonazza Can you verify the pr? I think it's ready to be merged.

File: modules/mbedtls_curl/http_client_curl.h
Comment (Faless) [inline modules/mbedtls_curl/http_client_curl.h:54]: Is there any reason to use this `RequestContext` with all pointers instead of just passing the `HTTPClientCurl` itself to (i.e. `this`) to the callbacks?
If we go for the `RequestContext` approach (which could be also good for future-proofing), those members should not be pointers, but the actual types.
We can add a `RequestContext current_context` member to `HTTPClientCurl` and pass the pointer to that to the callbacks (and add a `RequestContext::clear` for clearing it).

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (Faless) [inline modules/mbedtls_curl/http_client_curl.cpp:288]: Here, we don't want to set the ip address as the `CURLOPT_URL` but the actual hostname (this is what will be used for the `Host: hostname:port` header).

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (Faless) [inline modules/mbedtls_curl/http_client_curl.cpp:169]: This is blocking (`1000` is the timeout). Should probably be:
```
curl_multi_wait(curl, nullptr, 0, 0, nullptr)
```

And I think we can use this to implement the blocking mode.

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (Faless) [inline modules/mbedtls_curl/http_client_curl.cpp:431]: So the big problem with this is that between one `_poll_curl()` and the other there might not have been any new data received, which will be interpreted as the request being over.

**Review by Faless — CHANGES_REQUESTED** [review]
This looks very promising, I did spend a few hours hacking around it, testing various URLs, downloads, etc.
The biggest problem I have now is that I haven't found a reliable way to detect when curl has finished parsing the response (see my comment on the `STATUS_CONNECTED/DISCONNECTED` change).

We need to be able to do that for it to work properly in all cases (chunked/keep-alive/read-till-eof).

File: modules/mbedtls_curl/register_types.cpp
Comment (Faless) [inline modules/mbedtls_curl/register_types.cpp:45]: Forgot about this,
This does not need be registered, here, you would just call `HTTPClientCurl::make_default()` so API compatibility with and without the module is preserved.

**Comment by jonbonazza** [comment]
Thanks for the review @Faless and thank you soooooo much for the assistance with the build stuff @fire ! I unfortunately probably won't be able to get to addressing the comments tonight like I wanted to, but I'm confident that I'll be able to knock them out this week. 🤞 I left some follow up comments/questions on some of the concerns.

File: modules/mbedtls_curl/http_client_curl.h
Comment (jonbonazza) [reply modules/mbedtls_curl/http_client_curl.h]: Okay @Faless , I've updated the PR to remove the `RequestContext` struct entirely. It not stores the request state in the HttpClient object itself. I remember now that the reason I implemented the RequestContext was because I wasn't sure whether or not there could be multiple requests for the same client in flight at once. It doesn't seem like there can be given the way the interface is designed.

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (jonbonazza) [reply modules/mbedtls_curl/http_client_curl.cpp]: Okay this is fixed. And it also makes `http://www.example.com` finally work as well!

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (jonbonazza) [reply modules/mbedtls_curl/http_client_curl.cpp]: @Faless okay this is done. It seems to work as expected.

File: modules/mbedtls_curl/register_types.cpp
Comment (jonbonazza) [reply modules/mbedtls_curl/register_types.cpp]: Done.

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (jonbonazza) [reply modules/mbedtls_curl/http_client_curl.cpp:431]: Alright @Faless I'm not terribly happy with my solution to this, but it's the best I could think of after several hours of racking my brain. I left a comment in the code describing the problem and the up- and downsides to my solution.

File: modules/mbedtls_curl/http_client_curl.cpp
Comment (fire) [inline modules/mbedtls_curl/http_client_curl.cpp:326]: Was this supposed to be 2?

Comment (fire) [reply modules/mbedtls_curl/http_client_curl.cpp:326]: The diff was confusing. This is correct to not verify host. 

Can you rebase to like 2-3 commits?

Comment (jonbonazza) [reply modules/mbedtls_curl/http_client_curl.cpp:326]: Yep. Ive actually already done so locally. I was just waiting for CI to finish before I pushed. Ill do so shortly.

**Comment by jonbonazza** [comment]
@fire I've reduced the commits down to 2.

**Comment by Faless** [comment]
It doesn't so handle HEAD requests correctly:

```
@onready
var req = $HTTPRequest

# Called when the node enters the scene tree for the first time.
func _ready():
        req.connect("request_completed", Callable(self, "_completed"))
        # Stalls instead of completing after the header like the current implementation does.
        req.request("https://google.com/", [], false, HTTPClient.METHOD_HEAD)

```

[proj_httprequest_head.zip](https://github.com/godotengine/godot/files/7857018/proj_httprequest_head.zip)


Please refer to https://github.com/Faless/gd-tests/ (quite outdated) for some test URL.

Test at least:
- [ ] GET/HEAD/POST both SSL and not, both with known content length and without.
- [ ] Chunked/Content-Length keep-alive.
- [ ] Read-till-EOF.

Would be better to also test:
- [ ] PUT/DELETE (very common in rest APIs)
- [ ] OPTIONS (not sure, but should be supported by the current implementation, for CORS pre-flight).

**Comment by jonbonazza** [comment]
@Faless okay, I've tested every HTTP method I could think of and both chunked and not chunked for the ones that made sense. I also went through all the tests in your test project and made sure my responses matched what the tests expected. I also added all of the default headers that the old http client was adding.

One thing is I did screw up the git history so I ended up just squashing everything down.

**Comment by jonbonazza** [comment]
FWIW, a suuuper handy tool for doing these tests was https://httpbin.org

**Comment by mhilbrunner** [comment]
This is super not useful but I just want to chime in and say that its awesome how consistently you've been working away on this. Thanks for the efforts!

**Comment by jonbonazza** [comment]
More bugs were founds. Squashing. 🐞

**Comment by jonbonazza** [comment]
Ooookay. Squashed all the bugs I found. @fire the issue you found should also be resolved.

**Comment by fire** [comment]
I can still see a bug. Having trouble reducing a test case.

**Comment by int-72h** [comment]
This feature looks great (as at the moment I'm having to use a GDNative extension), will it be merged anytime soon?

**Comment by andy5995** [comment]
> This feature looks great (as at the moment I'm having to use a GDNative extension), will it be merged anytime soon?

Looks like it has merge conflicts at the moment.

**Pull request closed without merging**
